src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +563,544 @@ void ChunkScrub::chunk_scrub_common()
+    bool processed = false;
+  };
+  std::set<std::string> oid_for_evict;
+  static std::unordered_map<std::string, fp_store_entry_t> instant_fingerprint_store;
+  static std::shared_mutex fingerprint_lock;
+  static std::unordered_set<std::string> flushed_objects;
+  static std::shared_mutex flushed_lock;
+  size_t chunk_size;
+  fp_type_t fp_type;
+};
+
+SampleDedup::crawl_mode_t default_crawl_mode = SampleDedup::crawl_mode_t::DEEP;
+
+std::unordered_map<std::string, SampleDedup::fp_store_entry_t> SampleDedup::instant_fingerprint_store;
+std::shared_mutex SampleDedup::fingerprint_lock;
+std::unordered_set<std::string> SampleDedup::flushed_objects;
+std::shared_mutex SampleDedup::flushed_lock;
+
+SampleDedup::fp_type_t SampleDedup::get_fp_type(string fp_algo) {
+  if (fp_algo == "sha1") {
+    return fp_type_t::SHA1;
+  } else if (fp_algo == "sha256") {
+    return fp_type_t::SHA256;
+  } else if (fp_algo == "sha512") {
+    return fp_type_t::SHA512;
+  } else {
+    return fp_type_t::UNKNOWN;
+  }
+}
+
+void SampleDedup::crawl() {
+  try {
+    if (fp_type == fp_type_t::UNKNOWN) {
+      throw std::logic_error("unknown fingerprint algorithm");
+    }
+    prepare_rados();
+    ObjectCursor shard_start;
+    ObjectCursor shard_end;
+    std::tie(shard_start, shard_end) = get_shard_boundary();
+    cout << "new iteration thread: " << n <<std::endl;
+
+    for (ObjectCursor current_object = shard_start;
+        current_object < shard_end;) {
+      std::vector<ObjectItem> objects;
+      // Get the list of object IDs to deduplicate
+      std::tie(objects, current_object) = get_objects(
+          current_object,
+          shard_end,
+          100);
+
+      // Pick few objects to be processed. Crawling mode decides how many
+      // objects to pick (sampling ratio). Lower sampling ratio makes crawler
+      // have lower crawling overhead but find less duplication.
+      std::set<size_t> sampled_indexes = sample_object(objects.size());
+      for (size_t index : sampled_indexes) {
+        ObjectItem target = objects[index];
+        // Only process dirty objects which are expected not processed yet
+        if (is_dirty(target)) {
+          try_dedup_and_accumulate_result(target);
+        }
+      }
+    }
+
+    map<std::string,AioCompletion*> set_chunk_completions;
+    // Do set_chunk to make found duplicable chunks can be evicted by tier_evict()
+    for (auto& duplicable_chunk : duplicable_chunks) {
+      auto completion = set_chunk(duplicable_chunk);
+      if (completion != nullptr) {
+        set_chunk_completions[duplicable_chunk.oid] = completion;
+      }
+    }
+
+    vector<AioCompletion*> evict_completions;
+    for (auto& oid : oid_for_evict) {
+      auto completion_iter = set_chunk_completions.find(oid);
+      // Related set_chunk should be completed before tier_evict because
+      // tier_evict() only evict data processed by set_chunk() or tier_flush()
+      if (completion_iter != set_chunk_completions.end()) {
```

Comment:
I don't understand this part.

Probably,
aio_completion = async_set_chunk*(
wait_completion(vector<AioCompletion*>);
aio_completion = async_evict()
wait_completion(vector<AioCompletion*>);
?